gnome: Split header and code targets in gdbus_codegen()

inigomartinez opened the pull request:
The development version of `glib` (2.55.2) has acquired support for [generating gdbus header and source code files separately](https://bugzilla.gnome.org/show_bug.cgi?id=791015).

This PR allows dependencies to be more fine grained on those targets depending only on the header, which is needed for [some](https://bugzilla.gnome.org/show_bug.cgi?id=789877#c16) [ports](https://github.com/mesonbuild/meson/issues/2893#issuecomment-355856606).

inigomartinez commented:
Some side notes about this PR. I've tried to maintain backwards compatibility so, if `gdbus-codegen` is older than `2.55.2`, then an opaque object would be returned which represents to both `source` and `header`.

If it's a newer version, _three_ targets will be built: `source`, `header` and `docbook`. However, `docbook` output is not _real_, because the number of generated files is unknown, because it depends on the number of interfaces defined in the `xml` file.

jpakkane commented:
How recent is 2.55.2? It seems to be a bit of a bad form UI wise to change the output based on a version of a tool on the user's system.

xclaesse commented:
2.55.2 is current master, not released.

inigomartinez commented:
Yes it is still unreleased, but the changes are in place. The idea is to have meson support for whenever the new version is released. When used with older glib versions meson's output will still be the same.

`gdbus-codegen` is backwards compatible, after the modifications. It still has the same options plus some new ones, but all the programs can expect the same behaviour.

These new options allow to generate separately C source code and header, which wasn't possible before and it helps with dependencies.

inigomartinez commented:
I have reread again your comment and I've understood better what you meant.

The reason why the older implementation returned an opaque object was that `gdbus-codegen` wasn't able to generate C source code and header separately.

This new approach generates two different targets, one for the C source code and another for the header, by using the same command twice to bypass the original limitation. It will also maintain backward compatibility because this new array can also be seen as an opaque object.

jpakkane commented:
The only worry I have is that changing the return value has (probably very small) chance of breaking existing projects. Has someone tested building existing projects with this new code and did it work out fine?

And while we're on this, the code should check that the `docbook` kwarg is a string. Now it just uses it without validation.

nirbheek commented:
> The only worry I have is that changing the return value has (probably very small) chance of breaking existing projects

We should create a buildbot that builds all the projects on the Usage page once a day. That way we can catch issues early, write tests for them, and fix them. The current situation does not scale very well. The 0.44 release was very buggy despite our best efforts at testing.

@ignatenkobrain do you think we can do this somehow with Fedora resources?

inigomartinez commented:
> The only worry I have is that changing the return value has (probably very small) chance of breaking existing projects. Has someone tested building existing projects with this new code and did it work out fine?
>
> And while we're on this, the code should check that the docbook kwarg is a string. Now it just uses it without validation.

I have added support for `docbook` validation, and I have tested the software with some `freedesktop`/`gnome` packages that I have ported that are using `gdbus_codegen`:

* bijiben
* gnome-bluetooth
* gnome-calendar
* gnome-control-center
* gnome-online-accounts (not merged)
* gnome-photos (not merged)
* gnome-session
* gnome-settings-daemon
* gnome-terminal (not merged)
* gvfs
* NetworkManager

None of them had any issue with these changes so far. I have also tried these changes with these packages:

* glib
* gnome-shell

These packages didn't have any problems either. `gnome-shell` is an interesting package because its use case is not covered by meson's `gdbus-codegen` implementation: [it does generate `docbook` files without generating any source](https://gitlab.gnome.org/GNOME/gnome-shell/blob/master/docs/reference/shell/meson.build#L25).

> We should create a buildbot that builds all the projects on the Usage page once a day. That way we can catch issues early, write tests for them, and fix them. The current situation does not scale very well. The 0.44 release was very buggy despite our best efforts at testing.

Although its goal is totally different, [gnome-continuous](http://build.gnome.org) is building every updated gnome package with a development meson version (I don't know exactly if it's the last one). Just in case it helps.

nirbheek commented:
I think this testing is sufficient, we should get this in.

> Although its goal is totally different, gnome-continuous is building every updated gnome package with a development meson version (I don't know exactly if it's the last one). Just in case it helps.

Yes, I asked for that (I think) but GNOME use-cases are usually more well-covered than other projects simply because of how closely GNOME developers work with us, so we need a buildbot for the other projects too.

inigomartinez commented:
I've restarted the work by making some more changes to ease the way to also include the [gnome-shell](https://gitlab.gnome.org/GNOME/gnome-shell/blob/master/docs/reference/shell/meson.build#L25) case.

One of the first thing I noticed was that the file/target creation was not consistent between versions ([h, c, docbook] with the new glib code, [c, h, docbook] with the old glib code) and also between tools because `mkenums` and `genmarshal`are returning targets as [c, h].

I have more code locally and I need more testing, so please don't merge this yet.

ignatenkobrain commented on mesonbuild/modules/gnome.py:
```diff
@@ -900,21 +898,83 @@ def gdbus_codegen(self, state, args, kwargs):
+            output = namebase + '.c'
+            custom_kwargs = {'input': xml_file,
+                             'output': output,
+                             'command': cmd + ['--body', '--output', '@OUTDIR@/' +  output, '@INPUT@'],
```
__[Flake8]__

[E222] multiple spaces after operator


<!-- Comment by [SideCI](https://sideci.com) -->

inigomartinez commented:
I can't think of any way in which the backwards compatibility can't be broken.

The best way I can think of to maintain this compatibility is to create a new function and deprecate the docbook generation as it is now.

I think it is best to leave the changes in this PR as they are now, as they solve the original issue (the need to generate different targets for the header and code) and open a new PR for proper docbook generation.

So if you agree with these changes, they can be merged.

nirbheek commented:
I'll review this, then we can merge it.

jpakkane commented:
What's the review status of this one?

jpakkane commented:
The docs say `0.45.0` but they should say `0.46.0` as this has been delayed. But other than that LGTM. This can be merged once @nirbheek finishes his review.

inigomartinez commented:
> The docs say 0.45.0 but they should say 0.46.0 as this has been delayed. But other than that LGTM. This can be merged once @nirbheek finishes his review.

I have rebased and updated the changes. It says now 0.46.0. Although one of the tests has failed (it's actually a time out error), it passes all the unit tests here. Let's wait for @nirbheek.

nirbheek commented on mesonbuild/modules/gnome.py:
```diff
@@ -898,21 +896,83 @@ def gdbus_codegen(self, state, args, kwargs):
+                                 'command': cmd + ['--generate-docbook', docbook, '@INPUT@'],
+                                 'build_by_default': build_by_default
+                                 }
+                targets.append(build.CustomTarget(output, state.subdir, state.subproject, custom_kwargs))
```
I'm confused, the `else:` condition here (for older gdbus-codegen) is *also* returning a list of targets? Shouldn't this use the old mechanism of having a single custom target that generates the C source, header, and docbook?

Or did I misunderstand what you decided to do? Because right now the code contradicts the documentation.

inigomartinez replied:
This is due to a conversation at the IRC, where it was concluded that a version-dependent return was not a good design.

This, along with the fact that for previous versions it can be transparent, since the set of targets can be seen as the opaque target that was returned before, is the reason why I changed the implementation.

nirbheek replied:
So maybe we should have one custom target and return it thrice?

nirbheek replied:
Because by building things (potentially) thrice, there's a greater chance of things breaking. Returning the same target thrice will satisfy the API requirements and also greatly reduce the chances of regressions for existing projects with older gdbus-codegen.

inigomartinez replied:
> So maybe we should have one custom target and return it thrice?

The generation of a single target would not fit the goal of these changes, because separate targets are needed to handle dependencies on each target properly.

> Because by building things (potentially) thrice, there's a greater chance of things breaking. Returning the same target thrice will satisfy the API requirements and also greatly reduce the chances of regressions for existing projects with older gdbus-codegen.

Or do you mean splitting this new implementation into three different new functions for each target?

inigomartinez commented:
Here goes an update following your comments @nirbheek. Thanks!

I have also stashed all the previous commits into just two commits. I hope it's easier to review now.

inigomartinez commented on mesonbuild/modules/gnome.py:
```diff
@@ -898,21 +897,58 @@ def gdbus_codegen(self, state, args, kwargs):
+                             'command': cmd,
+                             'build_by_default': build_by_default
+                             }
+            targets = [build.CustomTarget(target_name, state.subdir, state.subproject, custom_kwargs)] * 3
```
Sure!

jpakkane commented:
Except that test failures seem relevant. :(

inigomartinez commented:
> Except that test failures seem relevant. :(

Right! there were a couple of little mistakes that are already fixed.
